### PR TITLE
Enable intptrcast for explicit casts

### PR DESCRIPTION
**Pull request by pvdrz**
I checked locally that this does not break miri on master. r? @RalfJung

### CODE REVIEW

File: src/librustc_mir/interpret/cast.rs [hunk 1]
```diff
@@ -239,17 +239,30 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpretCx<'mir, 'tcx, M> {
...
+            Int(IntTy::Isize) | Uint(UintTy::Usize) => {
+                if let Ok(bits) = self.force_bits(Scalar::Ptr(ptr), self.memory.pointer_size()) {
+                    self.cast_from_int(bits, src_layout, dest_layout)
+                } else {
+                    Ok(ptr.into())
+                }
+            }
+            Int(_) | Uint(_) => {
+                if let Ok(bits) = self.force_bits(Scalar::Ptr(ptr), self.memory.pointer_size()) {
+                    self.cast_from_int(bits, src_layout, dest_layout)
```
Comment (RalfJung) [inline src/librustc_mir/interpret/cast.rs]: You can share this first part with the branch above.
Then in the `else`, you can do something like `if dest_layout.size == memory.pointer_size() { Ok(ptr.into()) } else { ... }`.

Comment (RalfJung) [reply src/librustc_mir/interpret/cast.rs]: Oh and actually, you should just forward the error, not make up a new one.
```rust
match self.force_bits(...) {
Ok(bits) => ...
Err(_) if dest_layout.size == memory.pointer_size() => Ok(ptr.into()), // continue with ptr value at int type
Err(e) => Err(e), // propagate error
}
```

Comment (christianpoveda) [reply src/librustc_mir/interpret/cast.rs]: Even if the error I'm making is the one that was used in the original code?

Comment (RalfJung) [reply src/librustc_mir/interpret/cast.rs]: Better not to make assumptions like that. The intention here is to "catch the error if we can do better" (that's the `Ok(ptr)` branch). If we can't to better, no reason to catch and re-throw a new one.

**Comment by RalfJung** [comment]
This looks great, thanks! :)

@bors r+ rollup

**Comment by RalfJung** [comment]
While we are debugging why new Miri tests don't behave the way they should...

@bors r-

File: src/librustc_mir/interpret/cast.rs [hunk 2]
```diff
@@ -239,17 +239,25 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpretCx<'mir, 'tcx, M> {
...
+            Int(_) | Uint(_) => {
+                let size = self.memory.pointer_size();
+
+                match self.force_bits(Scalar::Ptr(ptr), size) {
+                    Ok(bits) => self.cast_from_int(bits, src_layout, dest_layout),
+                    Err(_) if dest_layout.size == size => Ok(ptr.into()),
+                    Err(e) => Err(e),
+                }
+            }
+            _ => return err!(Unimplemented(format!("ptr to {:?} cast", dest_layout.ty))),
```
Comment (RalfJung) [inline src/librustc_mir/interpret/cast.rs]:
```suggestion
            _ => bug!("invalid MIR: ptr to {:?} cast", dest_layout.ty),
```
While we are at it, we can turn this into a proper assertion. That should also clarify my point in the Miri PR. ;)

**Comment by christianpoveda** [comment]
Apparently I was running the tests with an older rustc binary. I recompiled and the tests had no regressions

**Comment by RalfJung** [comment]
Awesome.

@bors r+

**Pull request merged**